zjit/src/hir.rs

```diff
@@ -1154,6 +1161,7 @@ impl Function {
             Insn::ToNewArray { .. } => types::ArrayExact,
             Insn::ToArray { .. } => types::ArrayExact,
             Insn::ObjToString { .. } => types::BasicObject,
+            Insn::AnyToString { .. } => types::StringExact,
```

Comment:
This is not guaranteed to be `StringExact`, unfortunately; if `str` is a subclass of `String`, `str` gets returned unmodified. We can add a conditional here based on the inferred type of `str` or figure out another way to split the IR, but otherwise the type must be `types::String`

Reply:
you're right! good catch, thank you